staging/src/k8s.io/apimachinery/pkg/util/managedfields/extract.go

```diff
@@ -35,7 +35,11 @@ import (
 // that no managed fields were found for the fieldManager because other field managers
 // have taken ownership of all the fields previously owned by the fieldManager. It is
 // also possible the fieldManager never owned fields.
-func ExtractInto(object runtime.Object, objectType typed.ParseableType, fieldManager string, applyConfiguration interface{}) error {
+// Supported subresource values: "" (for spec), "status".
+func ExtractInto(object runtime.Object, objectType typed.ParseableType, fieldManager string, applyConfiguration interface{}, subresource string) error {
+	if len(subresource) > 0 && (subresource != "status") {
```

Comment:
The code using this doesn't appear to restrict to "status".  Why restrict it here?

Reply:
>> The code using this doesn't appear to restrict to "status". Why restrict it here?

> If this isn't a fundamental restriction, I'd prefer to have the underlying library flexible and the
> usage in the generated client only call with "status" if it wishes.

+@apelisse

I see your point.

I dug into this a bit with "scale" to make sure I understand how everything works. It's a bit-non obvious so wanted to summarize:
- `ExtractInto(<root object>, …, "status")` is a bit of a special case because "status" happens to be inlined in the root object, so it can be directly extracted from the root object with this invocation.
- "scale" is ALMOST inlined in the root resource; close but no cigar. `spec.replicas` and `status.replicas` are inlined at the exact same locations, but `status.selector` is in "scale" but maps to a different field in the root object (`spec.selector`) that is structured differently (`metav1.LabelSelector` instead of a `string`). 
- We don't have managed fields on subresources, only the root resource. I.e. if you GET the scale subresource, you get the object but it does not contain any `managedFields`, and if you GET the root resource, you get list of managedFields with a one that has `subresource=scale`, and the field path for the selector is `spec.selector` since that is where the field is located in the root object.
- This means that `ExtractInto(<root object>, …, "scale")` would return a filtered view of the root object with only the scale related fields, but could only be applied on the root object, not on the "scale" subresource.
- It also means that calling `ExtractInto(<scale object>, ..., "scale")` is not possible. If there was a subresource where all the fields are inlined in the main object at the exact same locations then it would work, but for scale it doesn't.
- We're currently only adding apply subresource support to "status" and "scale" (https://github.com/kubernetes/kubernetes/pull/98377).

I'm leaning toward keeping the check in place as a guard rail to developers, and either removing it or loosening it up once we support other subresources. Although I can see there is a good argument to be made that this check doesn't really belong at this level.

What do others think?

Reply:
Thanks @deads2k and @apelisse for the offline discussion.

The conclusion is: While ExtractInto is only useful for the "status" subresource for the existing builtin types, it could be useful with other subresources if extending Kubernetes, so we won't restrict what subresources can be passed in to it.

I've added a detailed section to the ExtractInto godoc explaining all this.